lib/dash/content_protection.js

```diff
@@ -262,6 +262,194 @@ shaka.dash.ContentProtection.parseFromRepresentation = function(
 };
 
 
+/**
+ * Gets a Widevine license URL from a content protection element
+ * containing a custom `ms:laurl` element
+ *
+ * @param {shaka.dash.ContentProtection.Element} element
+ * @return {string}
+ */
+shaka.dash.ContentProtection.getWidevineLicenseUrl = function(element) {
+  const mslaurlNode = shaka.util.XmlUtils.findChildNS(
+    element.node, 'urn:microsoft', 'laurl');
+  if (mslaurlNode) {
+    return mslaurlNode.getAttribute('licenseUrl') || '';
+  }
+  return '';
+};
+
+
+/**
+ * @typedef {{
+ *   type: number,
+ *   value: !Uint8Array
+ * }}
+ *
+ * @description
+ * The parsed result of a PlayReady object record.
+ *
+ * @property {number} type
+ *   Type of data stored in the record.
+ * @property {!Uint8Array} value
+ *   Record content.
+ */
+shaka.dash.ContentProtection.PlayReadyRecord;
+
+/**
+ * Enum for PlayReady record types.
+ * @enum {number}
+ */
+shaka.dash.ContentProtection.PLAYREADY_RECORD_TYPES = {
+  RIGHTS_MANAGEMENT: 0x001,
+  RESERVED: 0x002,
+  EMBEDDED_LICENSE: 0x003,
+};
+
+
+/**
+ * Parses an Array buffer starting at byteOffset for
+ * PlayReady Object Records. Each PRO Record is
+ * preceeded by it's PlayReady Record type and length
+ * in bytes.
+ *
+ * PlayReady Object Record format: https://goo.gl/FTcu46
+ *
+ * @param {!ArrayBuffer} recordData
+ * @param {number} byteOffset
+ * @return {!Array.<shaka.dash.ContentProtection.PlayReadyRecord>}
+ * @private
+ */
+shaka.dash.ContentProtection.parseMsProRecords_ = function(
+  recordData, byteOffset) {
+  const records = [];
+  const view = new DataView(recordData);
+
+  while (byteOffset < recordData.byteLength - 1) {
+    const type = view.getUint16(byteOffset, true);
+    byteOffset += 2;
+
+    const byteLength = view.getUint16(byteOffset, true);
+    byteOffset += 2;
+
+    goog.asserts.assert(
+      (byteLength & 1) === 0,
+      'expected byteLength to be an even number');
+
+    // Convert from bytes to WORD length
+    const begin = byteOffset / 2;
+    const end = (byteLength + byteOffset) / 2;
+
+    // subarray end is exclusive
+    const recordValue = new Uint16Array(recordData).subarray(begin, end);
+
+    records.push({
+      type: type,
+      value: new Uint8Array(recordValue),
```

Comment:
Couldn't you just do `new Uint8Array(recordData, byteOffset, byteLength)`?

Reply:
In that case, you should convert to a string here since the data will be UTF-16, not UTF-8 like the later string conversion expects.  Skipping every other byte like this only works for ASCII.

Reply:
You're right.  I was too focused on using the existing `parseXml` which was expecting UTF-8. Fixed.